signatures.md

```diff
@@ -10,205 +10,403 @@ We use signatures in filecoin to verify *something* was done by *someone*. For e
+	//   sk - a private key which cryptographically links `M` to `sig`
+	//
+	Sign(msg Message, addr Address, sk PrivateKey) (sig SignatureBytes, err error)
+}
+```
 
-Does not affect
+### Signature Interface
 
-- Transport encryption
-- File encodings (PoRep)
+```go
+type Signature interface {
 
-# Interface
+	// Aggregate generates a BLS signature aggregate from a set of BLS
+	// signatures. Associated `Address` must be passed to this function
+	// so that the aggregate can be verified. This function will pair a `sig`
+	// with an `Address` based on index in `sigs` and `addresses`.
+	//
+	// Out:
+	// 	 sig - a series of bytes representing a single BLS signature
+	//   err - a standard error message indicating any process issues
+	// In:
+	//   sigs - an array of SignatureBytes {sig_1, sig_2..., sig_n}
+	//   addrs - an array of Addresses {address_sig_1, address_sig_2..., address_sig_n}
+	//
+	Aggregate(sigs []SignatureBytes, addrs []Address) (sig SignatureBytes, err error)
+
+	// Verify validates the statement: only `M` could have generated `sig`
+	// given the validator has a message `m`, a signature `sig`, and a
+	// public key `pk`.
+	//
+	// Out:
+	//   valid - a boolean value indicating the signature is valid
+	//   err - a standard error message indicating any process issues
+	// In:
+	//   smsg - a SignedMessage which contains {Message, Address, SignatureBytes}
+	//   pk - a Public Key of the appropriate type
+	//
+	Verify(smsg SignedMessgage, pk PublicKey) (valid bool, err error)
+
+	// VerifyAggregate validates a collection of statements that are represeted by
+	// a signature aggregate: "For each M_i and PK_i in smsg, only this collection
+	// of {M}, {PK}, and signers could have generated `smsg.SignatureBytes`.
+	//
+	// Out:
+	//   valid - a boolean value indicating the signature is valid
+	//   err - a standard error message indicating any process issues
+	// In:
+	//   smsg - a SignedMessage which contains {[]Message, []Address, SignatureBytes}
+	//   pks - an array of Public Keys of the appropriate type associated with a message
+	// 	 at index `i` in []Message.
+	//
+	VerifyAggregate(smsg SignedMessgage, pks []PublicKeys) (valid bool, err error)
+
+	// Recover determines the public key associated with a particular signature. For
+	// ECDSA signatures public keys can be recovered from SignatureBytes and an
+	// associated message. For BLS signatures, public keys will be recovered by lookup
+	// using a mapping from Address to public key.
+	//
+	// Out:
+	//    pk - the public key associated with `M` who signed `msg`
+	//    err - a standard error message indicating any process issues
+	//    **
+	// In:
+	//    msg - a series of bytes representing the signed message
+	//    sig - a series of bytes representing a signature
+	//
+	Recover(msg Message, addr Address, sig SignatureBytes) (pk PublicKey, err error)
+  
+  //
+  //
+  Marshall(smsg SignedMessage) (SmsgBytes []byte, err error)
```

Comment:
@anorth: Request someone from go-filecoin help with `Marshall` and `Unmarshall`. I believe this should be CBOR encoding.

Reply:
I would be happy, nay happier(!) if these interfaces were pseudocode rather than Go code. That is to say, I think it would be fine and desirable for the spec to leave language-specifics to implementation, and instead just describe the functionality generally required primarily in prose (the comments above) and illustrative pseudocode. Would that decoupling make it easier for you?

The spec should certainly define the encoding, but it's sufficient to say "CBOR".

I realise that this may be a controversial position. If rejected, you could leave placeholders for the Go code here but you shouldn't be blocked on the go-filecoin team. Neither should this pre-empt that team's priorities and focus.

Reply:
I think if you're willing to commit people to working on this, I'm for it :-) 

If I recall, the reason we went with Go-style code blocks was b/c it was much faster for us to use at the time--since most of this work was first described in go-filecoin. I don't really have any objections to moving to a different structure as long as we realize that it'll be a lot of work for whoever has that undertaking.

Reply:
To be fully transparent, I am not intending to commit any people to working on it until the go-filecoin team needs to implement it. We can then backport form implementation to spec. If this doesn't work for the specs team, I'd suggest pseudocode.

Reply:
we use go-style pseudo code in the spec, it should not try to hard to be real working go code (i don’t want to see any if err blocks in the spec...). @bvohaska just do your best to write in pseudo code what should be done, ‘illustrative pseudocode’ as @anorth says.